SSL Behavior Changes

vtnerd opened the pull request:
The commit list explains the changes and why. I regret changing the names of `ssl_support_t` enum (which were unnecessarily redundant given the behavior of C++11 `enum class`), but it led to lots of lines changes.

Inspecting each commit manually might be preferable.

moneromooo-monero commented:
> There is no drop in security - a compromised exact match
> fingerprint has the same worse case failure.

Am I right in thinking that currently you can give any certificate, including one from random third parties ? If so, would this not decrease security by incentivizing people to offload their trust to that third party ?

vtnerd commented:
> Am I right in thinking that currently you can give any certificate, including one from random third parties ? If so, would this not decrease security by incentivizing people to offload their trust to that third party ?

I might be misunderstanding you - if the user specifies a file, only the certificates within that file are used for peer verification. The system "root CAs" are not loaded into the OpenSSL context. Using the built-in chain check means the user can create their own CA(s) without involving any outside party. The client just needs to specify the certificate file. The "users CA" can be a single self-signed certificate whose key is used by `monerod` directly.

Yes, you can use a 3rd party CA with this technique. I'm not sure how this would create an incentive to do so - perhaps in combination with the other changes? Perhaps that 3rd party wallets (i.e. not `monero-wallet-cli`) may recommend installing the CA file in the root store, making it _one of many_ CAs used for validation?

moneromooo-monero commented:
Yes, that's what I meant: if you accept "chained" certs, then you don't have to have the secret key to the "upper" cert, and people might start using a third party one if they have a couple personal cert that are signed by that third party one. If this happens, they suddenly gave access to anyone, no ?

vtnerd commented:
> if you accept "chained" certs, then you don't have to have the secret key to the "upper" cert

I took this excerpt because I wanted to make it clear that the user can generate their own "upper" certificate, and will have the secret key in that scenario. Given the second part of the sentence, I think the concern is over users getting a traditional SSL certificate signed, then have the client specify the root certificate instead of the locally generated certificate? And that the user wouldn't realize the ramifications of specifying the higher level certificate instead of the localized one?

I'm not sure what to make of this critique. The user would have to correctly identify and select a certificate not generated by the server operator while not realizing what it does. This scenario is possible, but is more work than using the certificate generated on the machine.

I think the reality is that the majority of users are going to use the default mode which does no verification at all.. The concern you raised could be handled by re-introducing a default  `set_verify_depth` value when specifying a custom certificate that can be manually overridden by advanced users - or no override at all. OpenSSL would still do all the certificate handling.

moneromooo-monero commented:
Yes. I understand you did this because a user can sign N certs with a single (owned) one, and I was indeed thinking that many users may just use a third party cert instead.
I'm not familiar with how these are used really, so it could well be that in practice it's an error few would do, if it requires technical competency in the first place. If you think this is the case, then fair enough.

vtnerd commented:
> I'm not familiar with how these are used really, so it could well be that in practice it's an error few would do, if it requires technical competency in the first place. If you think this is the case, then fair enough.

After your critique I can see how users may see "ca-certificate" and think they need to get a third party to sign it. That is the typical context for SSL - some outside service magically makes SSL work for the user. The user could incorrectly think that they need this magic third party and that specifying this third party certificate improves the security, when currently it does not. I can see users running `monerod` on VPS systems in particular setting this up incorrectly since "purchasing" a SSL certificate is integrated in those pesky sales carts.

I'll leave this as an open question for now. The real underlying issue is that there isn't a way for us to magically setup connections properly. The best solution I've come up with is a tutorial explaining how to set it up with fingerprints (beginner mode). Following the tutorial would result in optimal encryption + verification.

moneromooo-monero commented:
By "magically", you mean without having to whitelist a particular cert ?

vtnerd commented:
Yes. In my last comment, the referenced magic is the third party CA system used in browsers. Someone purchases a certificate for a domain name, and as-if-magic every web user is "secured". The user/client doesn't have to take any action at all other than "enabling https".

Purchasing a signed certificate for `monerod` isn't necessarily terrible. You can at least get the user/client to do `--daemon-ssl enable` or specify `--daemon-address https://example.com` (in some future patch I plan to auto-detect "https" in that switch and bump to SSL provided you don't do it first). With this series of patches enabling SSL with no whitelist is doing  the same checks a web browser would do - its not "awesome" but it is easy. Hopefully the user can be bothered to do a fingerprint or certificate config.

moneromooo-monero commented:
What needed the enums to change ? I don't see the cause.

moneromooo-monero commented:
After some discussion on IRC, I think it'd be nice to check whether a whitelisted cert is present in the OS cert repository, and warn about it if found. Also maybe add a --allow-derived-certificate or so option if wanting to use that system, which would protect people who don't realize they're allowing anyone allowed by a third party cert. What do you think ?

vtnerd commented:
Force pushed a rebase to merge in the http proxy patch. Title is now "do not merge" while I think about the default to "no chain" mode with user SSL certs. I backed out a patch that did this, it was fairly simple, but I'm not sure of the usefulness of a "no chain" default given OpenSSL behavior.

vtnerd commented:
> What needed the enums to change ? I don't see the cause.

The enums were named as-if they were **not** using C++11 scoped enums, so it was mainly an annonyance. I have since reverted the associated changes because they should not have been included in this PR. I also reverted changes that were only the removal of `epee::net_utils` namespacing.

> After some discussion on IRC, I think it'd be nice to check whether a whitelisted cert is present in the OS cert repository, and warn about it if found. Also maybe add a --allow-derived-certificate or so option if wanting to use that system, which would protect people who don't realize they're allowing anyone allowed by a third party cert. What do you think ?

There were two changes in the last force push, described in the first two paragraphs.

The user certificate now defaults to an allowed depth of "0" (chaining is disabled by default when the certificate is explicitly specified on the command line). `--daemon-ssl-allow-chained` overrides this default behavior. The server in `monerod` and `monero-wallet-rpc` does not allow chaining from client-specified certificates currently.

The fingerprint mode was also changed so that it **always** inspects the "top-level" server certificate. Previously if the server provided a chain, the client would check the fingerprint of the failing certificate, typically the root certificate.

As for an additional warning - nothing has been implemented yet. I would recommend that in a separate patch specifically for the wallet that is written to the CLI properly instead of the log. I haven't found a good way to determine if a user does `--daemon-ssl-allow-chained` and `--daemon-ssl-ca-certificates` with a system (third party) CA. The best recommendation I have is to warn if a wallet user is in autodetect or SSL "enabled" with no certificate or fingerprint specified. There is currently a warning, to the log but not necessarily CLI, if no certificate verification could be performed in autodetect mode (it "allows" the connection because a donwgrade attack is possible anyway).

vtnerd commented:
I also wanted to clarify in all of that - if the user specifies a certificate or fingerprint, the system (third party) CA files are **not** loaded. The difficulty is controlling/detecting if the user specifies the third party CA directly. I'm not certain how effective `--daemon-ssl-allow-chained` will be; in the worse case people just specify that flag because it stops the errors. A user specifying a third-party CA is arguably better than the standard SSL behavior of hundreds of root CAs.

moneromooo-monero commented on contrib/epee/include/net/net_ssl.h:
```diff
@@ -45,23 +46,90 @@ namespace net_utils
+        ca_path(),
+        auth(),
+        support(support),
+        verification(support == ssl_support_t::e_ssl_support_disabled ? ssl_verification_t::none : ssl_verification_t::system_ca)
```
The default goes to trusting any CA. Is that intended ?

vtnerd replied:
Yes. The choices are between no verification, the somewhat crummy system "root" CA system, or user provided information (when can be a onion/i2p address). I set this as the default, since this is "standard" SSL.

As per our discussion, if the user enables SSL via command line or the new `set_daemon` RPC, then a certificate, fingerprint or i2p/onion address must be provided.

fluffypony commented:
@vtnerd needs a rebase, please :)

vtnerd commented:
This was rebased on a few recent changes. The `set_daemon` flag now uses the "user" certificate CA file like the CLI changes. Additionally yet-another commit was added that prevents the usage of SSL system "root" CAs in the daemon connection from `monero-wallet-cli` or `monero-wallet-rpc`.

moneromooo-monero commented on src/wallet/wallet2.cpp:
```diff
@@ -357,22 +383,24 @@ std::unique_ptr<tools::wallet2> make_basic(const boost::program_options::variabl
-      !real_daemon.ends_with(".onion") && !real_daemon.ends_with(".i2p") &&
-      daemon_ssl_allowed_certificates.empty() && daemon_ssl_allowed_fingerprints.empty();
+    const bool verification_required =
+      ssl_options.support == epee::net_utils::ssl_support_t::e_ssl_support_enabled || use_proxy;
```
use_proxy will also be true if using a plaintext proxy, no ?

vtnerd replied:
The connection between wallet and proxy is always plaintext. This is attempting to force the usage of SSL through the proxy to verify the endpoint.
